File: apps/android/app/src/main/java/com/varn/app/Sample.kt
package com.varn.app

import android.content.Context
import org.json.JSONObject

// one entry of the shared sample set, which the manifest beside the lua files describes
data class Sample(val module: String, val label: String, val code: String) {
    val title: String get() = "$module · $label"
}

object Samples {
    private const val MANIFEST = "manifest.json"

    // the samples are assets shared with the other apps, so the list here is whatever the manifest declares
    fun load(context: Context): List<Sample> {
        val manifest = context.assets.open(MANIFEST).bufferedReader().use { it.readText() }
        val entries = JSONObject(manifest).getJSONArray("samples")

        return (0 until entries.length()).map { index ->
            val entry = entries.getJSONObject(index)
            val file = entry.getString("file")
            Sample(
                module = entry.getString("module"),
                label = entry.getString("label"),
                code = context.assets.open("lua/$file").bufferedReader().use { it.readText() },
            )
        }
    }
}

File: apps/ios/Sources/Sample.swift
import Foundation

// one entry of the shared sample set, which the manifest beside the lua files describes
struct Sample {
    let module: String
    let label: String
    let code: String

    var title: String { "\(module) · \(label)" }
}

enum Samples {
    private struct Manifest: Decodable {
        struct Entry: Decodable {
            let module: String
            let label: String
            let file: String
        }

        let samples: [Entry]
    }

    // the samples arrive as a folder reference, so they keep their directory inside the bundle
    static func load() -> [Sample] {
        guard let root = Bundle.main.url(forResource: "samples", withExtension: nil),
              let manifestData = try? Data(contentsOf: root.appendingPathComponent("manifest.json")),
              let manifest = try? JSONDecoder().decode(Manifest.self, from: manifestData)
        else {
            return []
        }

        return manifest.samples.compactMap { entry in
            let path = root.appendingPathComponent("lua").appendingPathComponent(entry.file)
            guard let code = try? String(contentsOf: path, encoding: .utf8) else {
                return nil
            }
            return Sample(module: entry.module, label: entry.label, code: code)
        }
    }
}
